server/service/osquery.go

```diff
@@ -1232,20 +1232,92 @@ func getFailingCalendarPolicies(policyResults map[uint]*bool, calendarPolicies [
 // We do this to not grow the main software queries and to ingest
 // all software together (one direct ingest function for all software).
 func preProcessSoftwareResults(
-	hostID uint,
+	host *fleet.Host,
 	results *fleet.OsqueryDistributedQueryResults,
 	statuses *map[string]fleet.OsqueryStatus,
 	messages *map[string]string,
 	overrides map[string]osquery_utils.DetailQuery,
 	logger log.Logger,
 ) {
+	//
 	vsCodeExtensionsExtraQuery := hostDetailQueryPrefix + "software_vscode_extensions"
-	preProcessSoftwareExtraResults(vsCodeExtensionsExtraQuery, hostID, results, statuses, messages, osquery_utils.DetailQuery{}, logger)
+	preProcessSoftwareExtraResults(vsCodeExtensionsExtraQuery, host.ID, results, statuses, messages, osquery_utils.DetailQuery{}, logger)
 
 	for name, query := range overrides {
 		fullQueryName := hostDetailQueryPrefix + "software_" + name
-		preProcessSoftwareExtraResults(fullQueryName, hostID, results, statuses, messages, query, logger)
+		preProcessSoftwareExtraResults(fullQueryName, host.ID, results, statuses, messages, query, logger)
+	}
+
+	// Filter out python packages that are also deb packages on ubuntu
+	pythonPackageFilter(host.Platform, results, statuses)
+}
+
+// pythonPackageFilter filters out duplicate python_packages that are installed under deb_packages on Ubuntu.
+// python_packages not matching a Debian package names are updated to "python3-packagename" to match OVAL definitions.
+func pythonPackageFilter(platform string, results *fleet.OsqueryDistributedQueryResults, statuses *map[string]fleet.OsqueryStatus) {
+	const pythonPrefix = "python3-"
+	const pythonSource = "python_packages"
+	const debSource = "deb_packages"
+	const linuxSoftware = hostDetailQueryPrefix + "software_linux"
+
+	// Return early if platform is not Ubuntu
+	// We may need to add more platforms in the future
+	if platform != "ubuntu" {
+		return
+	}
+
+	// Check if the 'software_linux' result and status
+	sw, ok := (*results)[linuxSoftware]
+	if !ok {
+		return
+	}
+
+	if status, ok := (*statuses)[linuxSoftware]; !ok || status != fleet.StatusOK {
+		return
+	}
+
+	// Extract the Python and Debian packages from the software list for filtering
+	// allocating space for 40 packages based on state of fresh ubuntu 24.04 install
+	pythonPackages := make([]string, 0, 40)
+	debPackages := make(map[string]struct{}, 40)
+	for _, row := range sw {
+		switch row["source"] {
+		case pythonSource:
+			loweredName := strings.ToLower(row["name"])
+			pythonPackages = append(pythonPackages, loweredName)
```

Comment:
The below loop through `sw` can be removed.

You can save the index to sw here -- convert `pythonPackages` to a map[string]int

Then, after this loop, loop through pythonPackages, and see if any of them need to be removed.

Sort indexes to remove in descending order, and remove them from `sw`

Reply:
thx for this, it did show some improvement with a minimal ~2 increase in mem allocations due to the append when "modifying" the results slice.

```
BenchmarkPreprocessUbuntuPythonPackageFilter-12    	   10000	    164683 ns/op	  174694 B/op	       8 allocs/op
```

I also realized a mistake with the original benchmark tests, as I was comparing this new filter against basically a no-op.  When I ran some manual benchmarks against the entire `ingestSoftware` func (~25ms), it was showing no impact as this filter is averaging ~0.15ms per operation.